srtp_writer_future.go

```diff
@@ -0,0 +1,95 @@
+// +build !js
+
+package webrtc
+
+import (
+	"io"
+	"sync/atomic"
+
+	"github.com/pion/rtp"
+	"github.com/pion/srtp"
+)
+
+// srtpWriterFuture blocks Read/Write calls until
+// the SRTP Session is available
+type srtpWriterFuture struct {
+	rtpSender      *RTPSender
+	rtcpReadStream atomic.Value // *srtp.ReadStreamSRTCP
+	rtpWriteStream atomic.Value // *srtp.WriteStreamSRTP
+}
+
+func (s *srtpWriterFuture) init() error {
```

Comment:
I would like to pass `context.Context` to all `Read/WriteRTP/RTCP` methods and handle cancel and timeout when blocking at here, as I proposed at https://github.com/pion/webrtc-v3-design/blob/01b2dee52ae4417a89c73385b8cae69b3877453b/rtpengine/io.go#L10-L35

Reply:
I forget the full story, but we attempted this before and it had serious performance issues. With that in mind I need to also confirm that interceptors don't cause any issues!

* https://github.com/pion/transport/pull/74
* https://github.com/pion/webrtc/pull/1266

Reply:
I am all for doing it eventually though, API wise it makes sense. I don't think we should do it for `/v3` though.

After this is merged and the Interceptors change I really want to feature freeze. Unless you think we are giving people a really bad experience

Reply:
If it blocks without context, I think it should have `SetDeadline` like `net.Conn` to make it able to be used safely. (`transport.Buffer` is doing so)

Reply:
@Sean-Der what do you think about https://github.com/pion/webrtc/pull/1546/files#r529263738

Reply:
@at-wat Oh nice, yes I am all for exposing that! We can add as an additional API. Really nice that it doesn't have to be a breaking change either.

Created https://github.com/pion/webrtc/issues/1553 that could be a nice first contribution for someone :)

Reply:
#1553 handles only the deadline of the underlying connection, and block due to this `init()` can't be handled.

I created a PR to add `ReadContext` to `packetio.Buffer` and the performance of the context version is better than the current master branch.
https://github.com/pion/transport/pull/104

Reply:
Is this PR ok to merge (mind approving?), since we can do the `context` stuff in another PR.

This PR is just moving around existing code, so we should have no behavior differences with `init`. All the function calls inside of it are non-blocking.

Reply:
Ah, right. Previous implementation was also blocking if Read is called without successful DTLS handshake.
https://github.com/pion/webrtc/blob/9ac8e9060faa876e14ca362b867e7596fa513691/rtpsender.go#L221-L226